### PR TITLE
feat: update toolchain on `lake update`

### LINKED ISSUE #2582: Lake update across lake versions
### Prerequisites

* [X] Put an X between the brackets on this line if you have done all of the following:
    * Check that your issue is not already [filed](https://github.com/leanprover/lean4/issues).
    * Reduce the issue to a minimal, self-contained, reproducible test case. Avoid dependencies to mathlib4 or std4.

### Description

Running `lake update` in an old project lead to errors because it tries to use and old version of lake to handle the modern lakefiles of dependencies.

### Context

This happens if you run `lake update` in a project that depend on Mathlib and is old enough to not know about `weakLeanArgs`.

### Steps to Reproduce

1. Create directory `foo` containing the lakefile
    ```
    import Lake
    open Lake DSL
    
    package «foo» {
      -- add any package configuration options here
    }
    
    require mathlib from git
      "https://github.com/leanprover-community/mathlib4.git"
    
    @[default_target]
    lean_lib «Foo» {
      -- add any library configuration options here
    }
    ```
    and the lean-toolchain file containing `leanprover/lean4:nightly-2023-06-20`

2. run `lake update`

**Expected behavior:** [Clear and concise description of what you expect to happen]

I expect lake to clone a recent version of mathlib without any error message.

**Actual behavior:** [Clear and concise description of what actually happens]

Get messages:
```
cloning https://github.com/leanprover-community/mathlib4.git to lake-packages/mathlib
error: ./lake-packages/mathlib/lakefile.lean:31:2: error: 'weakLeanArgs' is not a field of structure 'Lake.LeanLibConfig'
error: ./lake-packages/mathlib/lakefile.lean:55:2: error: 'weakLeanArgs' is not a field of structure 'Lake.LeanLibConfig'
error: ./lake-packages/mathlib/lakefile.lean: package configuration has errors
```

### Versions

Lean (version 4.0.0-nightly-2023-06-20, commit a44dd71ad62a, Release) on Ubuntu 22.04

I guess this cannot be fixed in old lake of course, but it would be nice to act now so that future lake won't break like this.

### LINKED ISSUE #2752: RFC: `lake update` should update the `lean-toolchain`
### Current behavior

`lake update` updates all `inherited: false` dependencies, even if doing so would move to a newer toolchain.
It does not change the value in `lean-toolchain`, not does it emit any diagnostic to indicate that this might need doing manually.

This frequently bites users of projects which depend on mathlib, as `lake update` gives them a mathlib version that does not compile with (or at least, has no cache for) their active toolchain

### Expected behavior

`lake update` should attempt to match the `lean-toolchain` of the new dependency versions. The design could be any of:
1. Look at all the versions of  `inherited: false` dependencies. If there is just one entry in this set, use it. If there is more than one, prompt the user.
2. Look at all the versions of  `inherited: false` dependencies. Attempt to pick the most recent of the options.
3. Add a `tracks_toolchain` key to the manifest, to indicate precisely which dependency the toolchain should be tracked from. For many users, this would be set to `mathlib4`.

### LINKED ISSUE #5615: No progress indicator for a long time when downloading a project in VSCode
When downloading a project using the VSCode extension, if that project is downstream of Mathlib there is a very long wait with no visual feedback or explanation of what is happening, while `lake` is cloning all of Mathlib.

This corresponds to the following terminal output:
```
/home/mhuisi/Lean/MathlibTest> lake exe cache get
<<<<<<< Very long pause here with no explanation of what is happening >>>>>>>>>>>>
info: batteries: cloning https://github.com/leanprover-community/batteries to '././.lake/packages/batteries'
info: Qq: cloning https://github.com/leanprover-community/quote4 to '././.lake/packages/Qq'
info: aesop: cloning https://github.com/leanprover-community/aesop to '././.lake/packages/aesop'
info: proofwidgets: cloning https://github.com/leanprover-community/ProofWidgets4 to '././.lake/packages/proofwidgets'
info: importGraph: cloning https://github.com/leanprover-community/import-graph to '././.lake/packages/importGraph'
info: LeanSearchClient: cloning https://github.com/leanprover-community/LeanSearchClient to '././.lake/packages/LeanSearchClient'
info: Cli: cloning https://github.com/leanprover/lean4-cli to '././.lake/packages/Cli'
```

Even though I know this is an issue, the cancel button or ctrl-c is very tempting because it feels like something is wrong.

@tydeu, my suggestion here is to *hard-code* a check for Mathlib (yes, I know you don't like the idea :-) and *before* cloning Mathlib print a "info: cloning Mathlib, depending on your connection this may take some time".

There needs to be something to tell the user something is happening.

An alternative solution would be to investigate doing a shallow clone of Mathlib, but I'm not sure how viable that is.

See [zulip](https://leanprover.zulipchat.com/#narrow/stream/113488-general/topic/gitpod.2Fcodespaces.2Flocal.20installation.20for.20Lean.20repo/near/474587581) discussion.

### PR DESCRIPTION
**Pull request by tydeu**
Lake will now update a package's `lean-toolchain` file on `lake update` if it finds the package's direct dependencies use a  newer compatible toolchain. To skip this step, use the `--keep-toolchain` CLI option.

Closes #2582. Closes #2752. Closes #5615.

### Toolchain update details

To determine "newest compatible" toolchain, Lake parses the toolchain listed in the packages' `lean-toolchain` files into four categories: release , nightly, PR, and other. For newness, release toolchains are compared by semantic version (e.g., `"v4.4.0" < "v4.8.0"` and `"v4.6.0-rc1" < "v4.6.0"`) and nightlies are compared by date (e.g., `"nightly-2024-01-10" < "nightly-2014-10-01"`). All other toolchain types and mixtures are incompatible. If there is not a single newest toolchain, Lake will print a warning and continue updating without changing the toolchain.

If Lake does find a new toolchain, Lake updates the workspace's `lean-toolchain` file accordingly and restarts the update process on the new Lake. If Elan is detected, it will spawn the new Lake process via `elan run`  with the same arguments Lake was initially run with. If Elan is missing, it will prompt the user to restart Lake manually and exit with a special error code (4).

### Other changes

To implement this new logic,  various other refactors were needed. Here are some key highlights:

* Logs emitted during package and workspace loading are now eagerly printed.
* The Elan executable used by Lake is now configurable by the `ELAN` environment variable.
* The `--lean` CLI option was removed. Use the `LEAN` environment variable instead.
* `Package.deps` / `Package.opaqueDeps` have been removed. Use `findPackage?` with a dependency's name instead.
* The dependency resolver now uses a pure breadth-first traversal to resolve dependencies. It also resolves dependencies in reverse order, which is done for consistency with targets. Latter targets shadow earlier ones and latter dependencies take precedence over earlier ones. **These changes mean the order of dependencies in a Lake manifest will change after the first `lake update` on this version of Lake.**

### CODE REVIEW

**Comment by tydeu** [comment]
!bench

**Comment by tydeu** [comment]
A Mathlib adaption for this PR is at leanprover-community/mathlib4#17953.

**Comment by Kha** [comment]
> The --lean CLI option was removed. The Lean setup is now configured solely through environment variables.

Could you please check this is compatible with https://github.com/leanprover-community/mathlib4/blob/master/scripts/bench/temci-config.run.yml#L10? It's been a while but AFAIR I tried using environment variables first but `--lean` worked best for this setup of proxy scripts.

File: src/lake/Lake/CLI/Error.lean [hunk 1]
```diff
@@ -7,6 +7,8 @@ Authors: Mac Malone
 namespace Lake
 open Lean (Name)
 
+deriving instance Repr for IO.Error
```
Comment (hargoniX) [inline src/lake/Lake/CLI/Error.lean:10]: I think deriving instances of built-in type classes for built-in types in `Lake` is not a good idea?

Comment (tydeu) [reply src/lake/Lake/CLI/Error.lean:10]: Lake does every so often need to define additional built-in instances for built-in types, and my understanding is the preferred procedure has been to define them in Lake (and then sometimes later upstream them to core). In this case, though, this is relic of some previous code, and thus I will  simply remove it.

**Pull request merged**